src/nvme/mi.c

```diff
@@ -0,0 +1,260 @@
+// SPDX-License-Identifier: LGPL-2.1-or-later
+/*
+ * This file is part of libnvme.
+ * Copyright (c) 2021 Code Construct Pty Ltd
+ *
+ * Authors: Jeremy Kerr <jk@codeconstruct.com.au>
+ */
+
+#include <errno.h>
+#include <stdlib.h>
+#include <stdlib.h>
+
+#include <ccan/endian/endian.h>
+
+#include "log.h"
+#include "mi.h"
+#include "private.h"
+
+/* MI-equivalent of nvme_create_root, but avoids clashing symbol names
+ * when linking against both libnvme and libnvme-mi.
+ */
+nvme_root_t nvme_mi_create_root(FILE *fp, int log_level)
```

Comment:
The question is if we should try to move common infrastructure and generic code out of `libnvme`? If I understood it correctly libnvme-mi can be used completely standalone too? I'd like to avoid having these function being duplicated. IIRC, there was already interests in splitting libnvme up into a generic part and a Linux specific one.

Reply:
I'm OK with a small amount of duplication here; the additional complexity of splitting libnvme doesn't seem to be a net win for this kind of thing. Having the MI interfaces as a "side-band" to the main interface seems OK for now.

Yes, libnvme-mi can be use completely standalone, without any local NVMe devices. The dependence on the kernel is just to create an `AF_MCTP` socket, but that could also be abstracted too (by replacing the MCTP backend with something else, as mentioned in https://github.com/linux-nvme/libnvme/pull/76#issuecomment-977326083 )

I do see one opportunity for unification though: we could be to make libnvme *completely* backend-independent; this would mean that the MI and ioctl-direct interfaces would use the same API, just a different constructor for the `nvme_root_t` - backed either by ioctls or MCTP (or something else for fabrics?); However, this would be a lot of churn, as there's lot of places in the current libnvme code that assume that everything is behind a file descriptor.  Overall, I don't think that's worth doing, particularly at this stage - I think we should give the MI interface some exposure to real applications to mature first.

Good point though - I'd definitely like to revisit this later, regardless of the decision now!

Reply:
I hear you. I also don't want to over engineering it. If it is just these few function you need for `libnvme-mi` than we might just go along with it, especially if it is a standalone library. Let's see if someone else has some thoughts on this.